persistence/data_models_json_fields.go

```diff
@@ -91,6 +92,127 @@ func (s *StateExecutionSequenceMapsJson) CompleteNewStateExecution(stateId strin
 	return nil
 }
 
+type StateExecutionWaitingQueuesJson struct {
+	// { queue_1: {(state_1, state_1_seq): waiting_count, (state_2, state_2_seq): waiting_count, ...} }
+	QueueToStatesMap map[string]map[string]int `json:"queueToStatesMap"`
+	// { (state_1, state_1_seq): total_waiting_count, (state_2, state_2_seq): total_waiting_count, ...} }
+	StateToQueueCountMap map[string]int `json:"stateToQueueCountMap"`
+	// [ message_1, message_2, ... ]
+	UnconsumedMessages []LocalQueueMessageInfoJson `json:"unconsumedMessages"`
+}
+
+func NewStateExecutionWaitingQueues() StateExecutionWaitingQueuesJson {
+	return StateExecutionWaitingQueuesJson{
+		QueueToStatesMap:     map[string]map[string]int{},
+		StateToQueueCountMap: map[string]int{},
+		UnconsumedMessages:   []LocalQueueMessageInfoJson{},
+	}
+}
+
+func (s *StateExecutionWaitingQueuesJson) ToBytes() ([]byte, error) {
+	return json.Marshal(s)
+}
+
+func NewStateExecutionWaitingQueuesFromBytes(bytes []byte) (StateExecutionWaitingQueuesJson, error) {
+	var waitingQueuesJson StateExecutionWaitingQueuesJson
+	err := json.Unmarshal(bytes, &waitingQueuesJson)
+	return waitingQueuesJson, err
+}
+
+func (s *StateExecutionWaitingQueuesJson) AddNewLocalQueueCommandForStateExecution(
+	stateExecutionId StateExecutionId, command xdbapi.LocalQueueCommand, anyOfCompletion bool) {
+	count := command.GetCount()
+	if count == 0 {
+		count = 1
+	}
+
+	stateExecutionIdKey := stateExecutionId.GetStateExecutionId()
+
+	m, ok := s.QueueToStatesMap[command.GetQueueName()]
+	if !ok {
+		m = map[string]int{}
+	}
+
+	if anyOfCompletion {
+		m[stateExecutionIdKey] = 1
+	} else {
+		m[stateExecutionIdKey] += int(count)
+	}
+	s.QueueToStatesMap[command.GetQueueName()] = m
+
+	if anyOfCompletion {
+		s.StateToQueueCountMap[stateExecutionIdKey] = 1
```

Comment:
I think we have some misunderstanding that I may not clearly explained to you earlier). 

The anyOf is on the command level, not on the command messages level. Say we have a commandRequest:
```
{
  WaitingType: AnyOf
  TimerCommands: [
     {TimerCommand1},
     {TimerCommand2}
   ],
   LocalQueueCommands:[
     {
          QueueName: "queue1",
          Count: 2,
     },
     {
          QueueName: "queue2",
          Count: 3,
     },
   ]
}
```
Then it will only complete waiting when any of two timer commands or local queue commands are completed. For the two local queue commands, each will require to get all the messages -- the first one will require to get 2 messages, and the second will require 3.